libs/java/gcp_zts_creds/src/main/java/com/yahoo/athenz/creds/gcp/GCPSIACredentials.java

```diff
@@ -0,0 +1,270 @@
+            httpConnection.setRequestMethod("GET");
+            httpConnection.setRequestProperty("Metadata-Flavor", "Google");
+            httpConnection.setConnectTimeout(ATTESTATION_CONNECT_TIMEOUT_MS);
+            httpConnection.setReadTimeout(ATTESTATION_READ_TIMEOUT_MS);
+            int status = httpConnection.getResponseCode();
+            String identityToken = new String(httpConnection.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+            if (status != 200) {
+                throw new Exception("Unable to generate GCF attestation data from URL \"" + gcpIdentityUrl + "\" : HTTP code " + status + " != 200");
+            }
+            // LOG.debug("GCF attestation-data: {}", identityToken);
+            return "{\"identityToken\":\"" + identityToken + "\"}";
+        } catch (IOException exception) {
+            throw new Exception("Unable to generate GCF attestation data from URL \"" + gcpIdentityUrl + "\" : ", exception);
+        } finally {
+            if (httpConnection != null) {
+                httpConnection.disconnect();
+            }
+        }
+    }
+
+    /** Build the Alternative DNS names (SAN's) */
+    private static GeneralName[] buildAlternativeDnsNames(
+            String athenzDomain,
+            String athenzService,
+            String gcpProjectId,
+            String certDomain) {
+        return new GeneralName[]{
+                new GeneralName(
+                        GeneralName.dNSName,
+                        new DERIA5String(athenzService + '.' + athenzDomain.replace('.', '-') + '.' + certDomain)),
+                new GeneralName(
+                        GeneralName.dNSName,
+                        new DERIA5String("gcf-" + gcpProjectId + '-' + athenzService + ".instanceid.athenz." + certDomain)),    // TODO: Not sure about this
+                new GeneralName(
+                        GeneralName.uniformResourceIdentifier,
+                        new DERIA5String("spiffe://" + athenzDomain + "/sa/" + athenzService)),
+        };
+    }
+
+    /** Request the Athenz certificate from ZTS server */
+    private static InstanceIdentity postInstanceRegisterInformation(
+            String athenzDomain,
+            String athenzService,
+            String athenzProvider,
+            String ztsUrl,
+            String attestationData,
+            String csr)
+            throws Exception {
+
+        // Construct an HTTP client.
+        RequestConfig config = RequestConfig.custom()
+                .setConnectTimeout(ZTS_CONNECT_TIMEOUT_MS)
+                .setSocketTimeout(ZTS_READ_TIMEOUT_MS)
+                .setRedirectsEnabled(false)
+                .build();
+        try (CloseableHttpClient httpClient = HttpClients.custom()
+                .setDefaultRequestConfig(config)
+                .build()) {
+
+            // Construct an HTTP POST request.
+            String postPayload = "{" +
```

Comment:
let's not generate the payload manually by generating the json here. We already have a model classes defined in athenz-zts-core.  

Let's define a com.yahoo.athenz.zts.InstanceRegisterInformation object, set the values in the object and then we can convert that object into json or I believe there are other Entity object setters.

Reply:
👍